### ISSUE TITLE
13-recommendation-systems-adult-census-income.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/alemantellini-machine-learning#8]
@alemantellini 

## Cumplimiento de Requisitos

| Requisito | Estado | Notas |
|-----------|--------|-------|
| Carga de datos | ✅ | URL remota + limpieza de espacios en columnas |
| Manejo de datos nulos | ✅ | Reemplazo '?' → NaN, imputación con moda para categóricas |
| Encoding categóricas | ⚠️ | LabelEncoder usado (pérdida de interpretabilidad vs one-hot) |
| Normalización numérica | ✅ | StandardScaler aplicado pre-split (⚠️ **data leakage!**) |
| Train/test split | ✅ | 80/20 con stratify=y — correcto |
| Comparación de modelos | ✅ | 4 modelos (LR, RF, GB, KNN) con validación cruzada |
| Métricas de evaluación | ✅ | Accuracy, AUC-ROC, cross-val score, confusion matrix |
| ROC curve visualization | ✅ | Gráfico de curva ROC incluido |
| Feature importance | ✅ | Top 10 mostrados para modelo árbol-based |
| Sistema recomendación | ✅ | Híbrido: simulación de cambios + k-NN content-based |
| Múltiples usuarios prueba | ✅ | 3 casos simulados con perfiles realistas |
| Análisis fairness | ✅ | Impacto por sexo y raza visualizado |
| Función interactiva | ✅ | `consultar_sistema()` permite entrada de usuario |
| Visualización final | ✅ | Resumen de 7 gráficos en una figura |
| Persistencia (modelos/scaler) | ❌ | **NO guardados** |
| Documentación | ✅ | Markdown bien estructurado, conclusiones reflexivas |

---

###  **CRÍTICO** (Must-Fix)

1. **Data leakage en StandardScaler**
   - `scaler.fit_transform(X_scaled_df)` hecho ANTES de `train_test_split()`
   - **Problema**: Test set "conoce" estadísticas de train set → métricas optimistas
   - **Fix**:
     ```python
     #  INCORRECTO (código actual):
     X_scaled = scaler.fit_transform(X)  # Fit en TODO X
     X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, ...)
     
     #  CORRECTO:
     X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
     scaler = StandardScaler()
     X_train_scaled = scaler.fit_transform(X_train)  # Fit SOLO en train
     X_test_scaled = scaler.transform(X_test)       # Transform test
     ```

2. ** Scaler y modelos NO guardados**
   - `mejor_modelo` entrenado pero **nunca guardado** con `joblib.dump()`
   - También `scaler`, `le_dict` (label encoders), `knn_content`
   - **Problema**: Imposible reproducir con nuevos usuarios
   - **Fix**:
     ```python
     import joblib
     joblib.dump(scaler, 'scaler_alessandra.pkl')
     joblib.dump(mejor_modelo, 'model_alessandra.pkl')
     joblib.dump(le_dict, 'label_encoders_alessandra.pkl')
     joblib.dump(knn_content, 'knn_content_alessandra.pkl')
     joblib.dump(all_features, 'feature_names_alessandra.pkl')
     print("✓ Modelo y preprocesamiento guardados")
     ```

3. **LabelEncoder crea multicolinealidad ordinal**
   - Educación codificada como 1-16 (sugiere orden jerárquico que modelo puede malinterpretar)
   - Ocupación, estado civil, etc. también sufren esto
   - **Problema**: Relaciones espurias (educación "5" no es necesariamente 5x education "1")
   - **Fix**: Usar OneHotEncoder con `drop='first'`:
     ```python
     from sklearn.preprocessing import OneHotEncoder
     encoder = OneHotEncoder(drop='first', sparse=False, handle_unknown='ignore')
     X_cat_encoded = encoder.fit_transform(X[features_cat])
     ```

5. **🟡 Función `predecir_ingreso()` puede error con LabelEncoder**
   - Si valor categórico no está en `le_dict[col].classes_`, salta a modo default
   - **Problema**: Predicciones inconsistentes para ocupaciones raras
   - **Fix**: Logging + fallback claro
